crates/bevy_pbr/src/extended_material.rs

```diff
@@ -0,0 +1,178 @@
+/// If the material `M` returns a non-default result from `fragment_shader()` it will be used in place of the standard
+/// fragment shader (`bevy_pbr::render::pbr_fragment.wgsl`). since all the standard material fields are
+/// present, the `pbr_fragment` shader function can be called from the custom shader (see
+/// the `extended_material` example).
+/// Alpha mode from the extended material is ignored, only the standard material's alpha mode
+/// is used.
+#[derive(Asset, Clone, TypePath)]
+pub struct ExtendedMaterial<M: Material> {
+    pub standard: StandardMaterial,
+    pub extended: M,
+}
+
+impl<M: Material> AsBindGroup for ExtendedMaterial<M> {
+    type Data = (
+        <StandardMaterial as AsBindGroup>::Data,
+        <M as AsBindGroup>::Data,
+    );
+
+    fn unprepared_bind_group(
+        &self,
+        layout: &BindGroupLayout,
+        render_device: &RenderDevice,
+        images: &RenderAssets<Image>,
+        fallback_image: &FallbackImage,
+    ) -> Result<bevy_render::render_resource::UnpreparedBindGroup<Self::Data>, AsBindGroupError>
+    {
+        // add together the bindings of the standard material and the user material
+        let UnpreparedBindGroup {
+            mut bindings,
+            data: standard_data,
+        } = StandardMaterial::unprepared_bind_group(
+            &self.standard,
+            layout,
+            render_device,
+            images,
+            fallback_image,
+        )?;
+        let extended_bindgroup = M::unprepared_bind_group(
+            &self.extended,
+            layout,
+            render_device,
+            images,
+            fallback_image,
+        )?;
+
+        bindings.extend(extended_bindgroup.bindings);
+
+        Ok(UnpreparedBindGroup {
+            bindings,
+            data: (standard_data, extended_bindgroup.data),
+        })
+    }
+
+    fn bind_group_layout_entries(
+        render_device: &RenderDevice,
+    ) -> Vec<bevy_render::render_resource::BindGroupLayoutEntry>
+    where
+        Self: Sized,
+    {
+        // add together the bindings of the standard material and the user material
+        let mut entries = StandardMaterial::bind_group_layout_entries(render_device);
+        entries.extend(M::bind_group_layout_entries(render_device));
+        entries
+    }
+}
+
+impl<M: Material> Material for ExtendedMaterial<M> {
+    fn vertex_shader() -> bevy_render::render_resource::ShaderRef {
+        match M::vertex_shader() {
+            bevy_render::render_resource::ShaderRef::Default => StandardMaterial::vertex_shader(),
+            specified => specified,
+        }
+    }
+
+    fn fragment_shader() -> bevy_render::render_resource::ShaderRef {
+        match M::fragment_shader() {
+            bevy_render::render_resource::ShaderRef::Default => StandardMaterial::fragment_shader(),
+            specified => specified,
+        }
+    }
+
+    fn alpha_mode(&self) -> crate::AlphaMode {
+        StandardMaterial::alpha_mode(&self.standard)
```

Comment:
I can think of a few scenarios where it would make sense to have the extended material control the `AlphaMode`, especially for effects that conditionally use `AlphaMode::Mask`, like screen door transparency for fading between different impostor levels, or for “despawning” or “dissolving” effects.

This is a relatively straightforward rule I think could work for enabling standard material to still control the alpha mode for simple extended materials, but still give extended material the ultimate control.

```suggestion
        let standard = StandardMaterial::alpha_mode(&self.standard);
        let extended = M::alpha_mode(&self.extended);

        if extended == crate::AlphaMode::Opaque {
            // When extended alpha mode is opaque, use standard's alpha mode
            standard
        } else {
            // Otherwise, extended will override standard
            extended
        }
```

<img src="https://github.com/bevyengine/bevy/assets/418473/6462e166-b7dd-436e-9ba5-e52e48a04bdf"  alt="screen door transparency">
<img src="https://github.com/bevyengine/bevy/assets/418473/9d3a9483-fd4d-42ae-8dfe-92d5d16eee63"  alt="dissolve effect">

Reply:
...whereas here it's not quite so obvious how to combine them. using opaque as default makes sense, but isn't entirely obvious.

i wonder if it would make sense to add another `MaterialExtension` trait that doesn't include depth bias or blend mode, and require that (`ExtendedMaterial<M: MaterialExtension>`)

Reply:
>`MaterialExtension` trait

i added this, i think it's better but let me know what you think.

Reply:
i just realised now that this doesn't give you control over the mode using the ExtendedMaterial - so it's no longer ambiguous but you'd still have to access the `StandardMaterial` asset in order to modify the blend mode. I definitely prefer only having one way to do it though, even if it means needing an extra asset lookup to modify the setting ...

Reply:
Independent of this PR, I've felt like AlphaMode is about RenderPhases and isn't specific to StandardMaterial. It should be possible to set in 2D as well.

Reply:
And yeah, having one way to configure alpha mode is much clearer imo.

Reply:
Hmm yeah if we eventually extract `AlphaMode` as its own component, then you can handle this with a system that just reads from whatever extended material you want and writes to `AlphaMode`, and it's both conceptually cleaner and fully consistent.